{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of pcHi-C data from human CD4+ T cells and ILC3 cells\n",
    "\n",
    "This notebook will walk through the analysis of pcHi-C data generated by Mikhail Spivakov's group. We will also analyze CD4+ alpha-beta T-cells in addition to the ILC3 data as a positive control sample."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "from sklearn.metrics import r2_score\n",
    "from scipy.stats import pearsonr\n",
    "from scipy.stats import spearmanr\n",
    "from decimal import Decimal\n",
    "import matplotlib.ticker as ticker\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ChicagoData(object):\n",
    "    \"\"\"Import CHiCAGO data\n",
    "    \"\"\"\n",
    "    def __init__(self,\n",
    "                 filename: str,\n",
    "                 drop_off_target_bait: bool = True,\n",
    "                 drop_off_target_oe: bool = True,\n",
    "                 drop_trans_chrom: bool = True,\n",
    "                 score_col: str = None,\n",
    "                 score_val: int = 5,\n",
    "                 remove_p2p: bool= True\n",
    "                 ):\n",
    "        \"\"\"Initialize the object\n",
    "\n",
    "        Args:\n",
    "            filename (str): Input CHICAGO txt file\n",
    "            drop_off_target (bool, optional): Drop off target interactions. Defaults to =True.\n",
    "            drop_trans_chrom (bool, optional): Drop transchromosomal interactions. Defaults to =True.\n",
    "        \"\"\"\n",
    "        # Set filename to the input filename\n",
    "        self.filename = filename\n",
    "        # Set whether to drop off target baits\n",
    "        self.drop_off_target_bait = drop_off_target_bait\n",
    "        # Set whether to drop off target oe\n",
    "        self.drop_off_target_oe = drop_off_target_oe\n",
    "        # Set whether to drop transchromosomal interactions\n",
    "        self.drop_trans_chrom = drop_trans_chrom\n",
    "        # Score column name for filtering\n",
    "        self.score_col = score_col\n",
    "        # Score value threshold\n",
    "        self.score_val = score_val\n",
    "        # Remove promoter to promoter interactions\n",
    "        self.remove_p2p = remove_p2p\n",
    "\n",
    "        # Read file into DF\n",
    "        self._read_file_()\n",
    "        \n",
    "        # Format the DF\n",
    "        self._format_file_()\n",
    "        \n",
    "        # Filter the formatted DF\n",
    "        self._filter_file_()\n",
    "        \n",
    "        # Get the PIR df\n",
    "        self._get_PIR_df_()\n",
    "        \n",
    "        # Get the bait df\n",
    "        self._get_bait_df_()\n",
    "        \n",
    "        # Get the combined df\n",
    "        self._get_combined_df_()    \n",
    "        \n",
    "    def _read_file_(self):\n",
    "        \"\"\"Read in original file\n",
    "        \"\"\"\n",
    "        # Read in original file and save\n",
    "        self.input_df =  pd.read_csv(self.filename, sep=\"\\t\", header=0, low_memory=False)\n",
    "    \n",
    "    def _format_file_(self):\n",
    "        \"\"\"Format CHICAGO file\n",
    "        \"\"\"\n",
    "        # Create a copy of the raw input to be manipulated\n",
    "        df = self.input_df.copy()\n",
    "        \n",
    "        # Format the chromosome names\n",
    "        df[\"baitChr\"] = \"chr\" + df[\"baitChr\"].apply(str)\n",
    "        df[\"oeChr\"] = \"chr\" + df[\"oeChr\"].apply(str)\n",
    "        \n",
    "        df[\"oe_interval_ID\"] = df[\"oeChr\"] + \":\" + \\\n",
    "                   df[\"oeStart\"].apply(str) + \"-\" + \\\n",
    "                   df[\"oeEnd\"].apply(str)\n",
    "\n",
    "        df[\"bait_interval_ID\"] = df[\"baitChr\"] + \":\" + \\\n",
    "                   df[\"baitStart\"].apply(str) + \"-\" + \\\n",
    "                   df[\"baitEnd\"].apply(str)\n",
    "\n",
    "        df[\"interaction_ID\"] = df[\"bait_interval_ID\"] + \"_\" + df[\"oe_interval_ID\"].apply(str)\n",
    "        \n",
    "        # Set the variable to the formatted df\n",
    "        self.bait_ID = df[\"baitID\"].unique()\n",
    "        \n",
    "        self.oe_ID = df[\"oeID\"].unique()\n",
    "\n",
    "        self.bait_interval_ID = df[\"bait_interval_ID\"].unique()\n",
    "        \n",
    "        self.oe_interval_ID = df[\"oe_interval_ID\"].unique()\n",
    "\n",
    "        self.df = df\n",
    "\n",
    "    def _filter_file_(self):\n",
    "        \"\"\"Filter the formatted CHICAGO results\n",
    "        \"\"\"\n",
    "        # Drop the off target baits\n",
    "        if self.drop_off_target_bait:\n",
    "            self.df[self.df[\"baitName\"] != \"off_target\"]\n",
    "\n",
    "        # Drop the off target OE names\n",
    "        if self.drop_off_target_oe:\n",
    "            self.df[self.df[\"oeName\"] != \"off_target\"]\n",
    "\n",
    "        # Drop the trans chromosomal interactions\n",
    "        if self.drop_trans_chrom:\n",
    "            self.df = self.df[self.df[\"dist\"] != \".\"]\n",
    "                        \n",
    "            self.df = self.df[self.df[\"dist\"].apply(float) != 0]\n",
    "\n",
    "            self.df = self.df.dropna(subset=[\"dist\"])\n",
    "\n",
    "            self.df = self.df[self.df[\"baitChr\"] == self.df[\"oeChr\"]]\n",
    "\n",
    "        # Filter the specific score column by a specific value\n",
    "        if self.score_col:\n",
    "            self.df = self.df[self.df[self.score_col] >= self.score_val]\n",
    "            \n",
    "        # Drop promoter to promoter interactions\n",
    "        if self.remove_p2p:\n",
    "            self.df = self.df[self.df.oeName == \".\"]\n",
    "            \n",
    "            self.df = self.df[~self.df.oe_interval_ID.isin(self.bait_interval_ID)]\n",
    "        \n",
    "    def _get_PIR_df_(self):\n",
    "        \"\"\"Get a DF of all PIR interactions\n",
    "        \"\"\"\n",
    "        self.pir_df = self.df[[\"oeChr\", \"oeStart\", \"oeEnd\", \"oe_interval_ID\"]].drop_duplicates(subset=[\"oeChr\", \"oeStart\", \"oeEnd\"], keep=\"first\")\n",
    "        \n",
    "    def _get_bait_df_(self):\n",
    "        \"\"\"Get a DF of baits\n",
    "        \"\"\"\n",
    "        self.bait_df = self.df[[\"baitChr\", \"baitStart\", \"baitEnd\", \"bait_interval_ID\"]].drop_duplicates(subset=[\"baitChr\", \"baitStart\", \"baitEnd\"], keep=\"first\")\n",
    "        \n",
    "    def _get_combined_df_(self):\n",
    "        \"\"\"Get a comined DF\n",
    "        \"\"\"\n",
    "        tmp_df = self.pir_df.copy()\n",
    "        tmp_df2 = self.bait_df.copy()\n",
    "\n",
    "        tmp_df.columns = [\"Chr\", \"Start\", \"Stop\", \"ID\"]\n",
    "        tmp_df2.columns = [\"Chr\", \"Start\", \"Stop\", \"ID\"]\n",
    "\n",
    "        self.unique_features = pd.concat([tmp_df, tmp_df2])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def counts_dict(input_filename: str):\n",
    "    \"\"\"Create a dictionary of Feature IDs to counts\n",
    "\n",
    "    Args:\n",
    "        input_filename (str): Input filename\n",
    "\n",
    "    Returns:\n",
    "        dict: Dictionary mapping Feature IDs to counts\n",
    "    \"\"\"\n",
    "    counts_df = pd.read_csv(input_filename, sep=\"\\t\", header=None, names=[\"chr\", \"start\", \"stop\", \"ID\", \"counts\"])\n",
    "            \n",
    "    counts_dict =  pd.Series(counts_df[\"counts\"].values,index=counts_df[\"ID\"]).to_dict()\n",
    "\n",
    "    return counts_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "class GeneExpressionAnalysis(object):\n",
    "    def __init__(self,\n",
    "                 df: pd.DataFrame,\n",
    "                feature_col: str,\n",
    "                mean_col: str,\n",
    "                title: str = \"\",\n",
    "                xlabel: str = \"\",\n",
    "                log: bool = True,\n",
    "                output_filename: str = \"\",\n",
    "                non_zero: bool = True, \n",
    "                drop_na: bool = True\n",
    "                ):\n",
    "        \"\"\"Gene Expression Analysis of pcHiC associated feature counts\n",
    "\n",
    "        Args:\n",
    "            df (pd.DataFrame): _description_\n",
    "            feature_col (str): _description_\n",
    "            mean_col (str): _description_\n",
    "            title (str): _description_\n",
    "            xlabel (str): _description_\n",
    "            log (bool): _description_\n",
    "            plot (bool): _description_\n",
    "            plot_filename (str): _description_\n",
    "        \"\"\"\n",
    "        self.df = df\n",
    "        self.feature_col = feature_col\n",
    "        self.mean_col = mean_col\n",
    "        self.log = log\n",
    "        self.output_filename = output_filename\n",
    "        self.nonzero = non_zero\n",
    "        self.dropna = drop_na\n",
    "        self.xlabel = xlabel\n",
    "        self.title = title\n",
    "        \n",
    "        # Filter the dataframe based on any specific criteria\n",
    "        self._filter_df_()\n",
    "        \n",
    "        # Calculate the spearman correlation of the the mean gene expression and the feature counts column\n",
    "        self._calculate_spearman_()\n",
    "        \n",
    "        self._get_PIR_count_v_mean_()\n",
    "        \n",
    "        #self.plot_expression_analysis()\n",
    "        \n",
    "    def _filter_df_(self):\n",
    "        if self.nonzero:\n",
    "            self.df = self.df[self.df[self.mean_col] > 0]\n",
    "\n",
    "        if self.dropna:\n",
    "            self.df = self.df[~self.df[self.feature_col].isna()]\n",
    "            self.df = self.df[~self.df[self.mean_col].isna()]\n",
    "\n",
    "        self.df[\"GeneName_MeanExpression\"] = self.df[\"GeneName\"] \\\n",
    "            + \" \" + self.df[self.mean_col].apply(str)\n",
    "            \n",
    "    def _calculate_spearman_(self):\n",
    "        tmp_df = self.df[[self.feature_col, self.mean_col]].copy()\n",
    "        \n",
    "        self.corr, self.pval = spearmanr(tmp_df[self.feature_col], tmp_df[self.mean_col])\n",
    "    \n",
    "    def _get_PIR_count_v_mean_(self):\n",
    "        \"\"\"Create the dataframe of the number of features overlapping PIRs by mean gene expression\n",
    "        \n",
    "        First you groupby the feature counts column. Then you find all of the mean gene expression values\n",
    "        associated with the number of features overlapping a PIR. The output is a dataframe that can be plotted\n",
    "        \"\"\"\n",
    "        # Create the dataframe\n",
    "        self.pir_count_v_mean = pd.DataFrame(self.df.groupby([self.feature_col])[\"GeneName_MeanExpression\"].apply(list)).reset_index().explode(\"GeneName_MeanExpression\")\n",
    "    \n",
    "        self.pir_count_v_mean[self.feature_col] = self.pir_count_v_mean[self.feature_col].apply(int)\n",
    "        \n",
    "        self.pir_count_v_mean[[\"Gene_Name\", \"Mean_Gene_Expression\"]] = self.pir_count_v_mean[\"GeneName_MeanExpression\"].str.split(\" \", n=2, expand=True)\n",
    "        \n",
    "        self.pir_count_v_mean.drop(\"GeneName_MeanExpression\",axis=1, inplace=True)\n",
    "        \n",
    "        self.pir_count_v_mean[\"Mean_Gene_Expression\"] = self.pir_count_v_mean[\"Mean_Gene_Expression\"].apply(float)\n",
    "        \n",
    "        self.pir_count_v_mean.to_csv(self.output_filename + \".tsv\", sep=\"\\t\", index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting PIRs From CHiCAGO output file\n",
    "\n",
    "I wrote a python class object to work with the CHiCAGO output text file. This piece of code will perform filtering of specific types of interactions, like promoter-to-promoter, or trans-chromosomal interactions. \n",
    "\n",
    "Example:\n",
    "\n",
    "```python\n",
    "input_file = \"../data/CHICAGO/hg38/inputs/ILC_5kb_within_newbmap_CHiCAGO_ABC_peakm.txt\"\n",
    "\n",
    "ILC3_data = ChicagoData(input_file)\n",
    "\n",
    "ILC3_data = ChicagoData(input_file, \n",
    "                        drop_off_target_bait=True, \n",
    "                        drop_off_target_oe=False, \n",
    "                        drop_trans_chrom=True,\n",
    "                        score_col=\"merged_score\",\n",
    "                        score_val=5,\n",
    "                        remove_p2p=True)\n",
    "                        \n",
    "ILC3_data.pir_df\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input Chicago File\n",
    "input_file = \"../data/CHICAGO/hg38/inputs/ILC_5kb_within_newbmap_CHiCAGO_ABC_peakm.txt.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "ILC3_data = ChicagoData(input_file, \n",
    "                        drop_off_target_bait=True, \n",
    "                        drop_off_target_oe=False, \n",
    "                        drop_trans_chrom=True,\n",
    "                        score_col=\"merged_score\",\n",
    "                        score_val=5,\n",
    "                        remove_p2p=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>oeChr</th>\n",
       "      <th>oeStart</th>\n",
       "      <th>oeEnd</th>\n",
       "      <th>oe_interval_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>11159</td>\n",
       "      <td>chr1:1-11159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>916863</td>\n",
       "      <td>923071</td>\n",
       "      <td>chr1:916863-923071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>939433</td>\n",
       "      <td>944716</td>\n",
       "      <td>chr1:939433-944716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>972542</td>\n",
       "      <td>978821</td>\n",
       "      <td>chr1:972542-978821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>989911</td>\n",
       "      <td>995125</td>\n",
       "      <td>chr1:989911-995125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84672</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13457330</td>\n",
       "      <td>13463420</td>\n",
       "      <td>chrY:13457330-13463420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84673</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13468452</td>\n",
       "      <td>13473885</td>\n",
       "      <td>chrY:13468452-13473885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84674</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13473886</td>\n",
       "      <td>13478002</td>\n",
       "      <td>chrY:13473886-13478002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84675</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13481458</td>\n",
       "      <td>13487154</td>\n",
       "      <td>chrY:13481458-13487154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84680</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13704104</td>\n",
       "      <td>13709537</td>\n",
       "      <td>chrY:13704104-13709537</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40782 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      oeChr   oeStart     oeEnd          oe_interval_ID\n",
       "0      chr1         1     11159            chr1:1-11159\n",
       "1      chr1    916863    923071      chr1:916863-923071\n",
       "4      chr1    939433    944716      chr1:939433-944716\n",
       "5      chr1    972542    978821      chr1:972542-978821\n",
       "8      chr1    989911    995125      chr1:989911-995125\n",
       "...     ...       ...       ...                     ...\n",
       "84672  chrY  13457330  13463420  chrY:13457330-13463420\n",
       "84673  chrY  13468452  13473885  chrY:13468452-13473885\n",
       "84674  chrY  13473886  13478002  chrY:13473886-13478002\n",
       "84675  chrY  13481458  13487154  chrY:13481458-13487154\n",
       "84680  chrY  13704104  13709537  chrY:13704104-13709537\n",
       "\n",
       "[40782 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ILC3_data.pir_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Processing Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/caz3so/workspaces/tacazares/pchic/notebooks'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "ilc3_file = \"../data/CHICAGO/hg38/inputs/ILC_5kb_within_newbmap_CHiCAGO_ABC_peakm.txt.gz\"\n",
    "cd4_file = \"../data/CHICAGO/hg38/inputs/CD4_1M_50K_5kb_within_newbmap_CHiCAGO_peakm.txt.gz\"\n",
    "cd4_ABC = \"../data/CHICAGO/hg38/inputs/CD4_1M_50K_5kb_within_newbmap_CHiCAGO_ABC_peakm.txt.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir = \"../data/CHICAGO/hg38/PIR\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import ILC3 data\n",
    "ilc3 = ChicagoData(ilc3_file, \n",
    "                        drop_off_target_bait=True, \n",
    "                        drop_off_target_oe=False, \n",
    "                        drop_trans_chrom=True,\n",
    "                        score_col=\"merged_score\",\n",
    "                        score_val=5,\n",
    "                        remove_p2p=True)\n",
    "\n",
    "# Import cd4 file CD4_1M_50K_5kb_within_newbmap_CHiCAGO_peakm.txt\n",
    "cd4 = ChicagoData(cd4_file, \n",
    "                        drop_off_target_bait=True, \n",
    "                        drop_off_target_oe=False, \n",
    "                        drop_trans_chrom=True,\n",
    "                        score_col=\"merged_score\",\n",
    "                        score_val=5,\n",
    "                        remove_p2p=True)\n",
    "\n",
    "# Import cd4_abc file CD4_1M_50K_5kb_within_newbmap_CHiCAGO_ABC_peakm.txt\n",
    "cd4_abc = ChicagoData(cd4_ABC, \n",
    "                        drop_off_target_bait=True, \n",
    "                        drop_off_target_oe=False, \n",
    "                        drop_trans_chrom=True,\n",
    "                        score_col=\"merged_score\",\n",
    "                        score_val=5,\n",
    "                        remove_p2p=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Write PIR to files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ilc3.pir_df.to_csv(os.path.join(out_dir, \"ILC_5kb_within_newbmap_CHiCAGO_ABC_peakm_PIR.bed.gz\"), sep=\"\\t\", header=False, index=False)\n",
    "#cd4.pir_df.to_csv(os.path.join(out_dir, \"CD4_1M_50K_5kb_within_newbmap_CHiCAGO_peakm_PIR.bed.gz\"), sep=\"\\t\", header=False, index=False)\n",
    "#cd4_abc.pir_df.to_csv(os.path.join(out_dir, \"CD4_1M_50K_5kb_within_newbmap_CHiCAGO_ABC_peakm_PIR.bed.gz\"), sep=\"\\t\", header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ilc3.unique_features.to_csv(os.path.join(out_dir, \"ILC_5kb_within_newbmap_CHiCAGO_ABC_peakm_PIRwPromoters.bed.gz\"), sep=\"\\t\", header=False, index=False)\n",
    "#cd4.unique_features.to_csv(os.path.join(out_dir, \"CD4_1M_50K_5kb_within_newbmap_CHiCAGO_peakm_PIRwPromoters.bed.gz\"), sep=\"\\t\", header=False, index=False)\n",
    "#cd4_abc.unique_features.to_csv(os.path.join(out_dir, \"CD4_1M_50K_5kb_within_newbmap_CHiCAGO_ABC_peakm_PIRwPromoters.bed.gz\"), sep=\"\\t\", header=False, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Expression analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Intersect PIR with genetic features\n",
    "\n",
    "I used `bedtools intersect -c` to count the number of features that overlapped PIR intervals. \n",
    "\n",
    "```bash\n",
    "for PIR_BED in /Users/caz3so/workspaces/tacazares/pchic/data/CHICAGO/hg38/PIR/CD4*bed.gz;\n",
    "    do\n",
    "        bedtools intersect -a ${PIR_BED} -b ../data/ATAC/CD4_ATAC_peaks.bed -c > ../data/PIR_overlap/`basename ${PIR_BED} .bed.gz`_overlapATAC.bed\n",
    "        bedtools intersect -a ${PIR_BED} -b ../data/CHIP/S008H1H1.ERX547940.H3K27ac.bwa.GRCh38.20150527.bed -c > ../data/PIR_overlap/`basename ${PIR_BED} .bed.gz`_overlapH3K27ac.bed\n",
    "        bedtools intersect -a ${PIR_BED} -b ../data/CHIP/S008H1H1.ERX547958.H3K4me3.bwa.GRCh38.20150527.bed -c > ../data/PIR_overlap/`basename ${PIR_BED} .bed.gz`_overlapH3K4me3.bed\n",
    "        bedtools intersect -a ${PIR_BED} -b ../data/RE/CD4_RE.bed -c > ../data/PIR_overlap/`basename ${PIR_BED} .bed.gz`_overlapRE.bed\n",
    "    done\n",
    "\n",
    "for PIR_BED in /Users/caz3so/workspaces/tacazares/pchic/data/CHICAGO/hg38/PIR/ILC*bed.gz;\n",
    "    do\n",
    "        bedtools intersect -a ${PIR_BED} -b ../data/ATAC/ILC3_ATAC_peaks.bed -c > ../data/PIR_overlap/`basename ${PIR_BED} .bed.gz`_overlapATAC.bed\n",
    "        bedtools intersect -a ${PIR_BED} -b ../data/CHIP/ILC3_H3K27ac_peaks.bed -c > ../data/PIR_overlap/`basename ${PIR_BED} .bed.gz`_overlapH3K27ac.bed\n",
    "        bedtools intersect -a ${PIR_BED} -b ../data/CHIP/ILC3_H3K4me3_peaks.bed -c > ../data/PIR_overlap/`basename ${PIR_BED} .bed.gz`_overlapH3K4me3.bed\n",
    "        bedtools intersect -a ${PIR_BED} -b ../data/RE/ILC3_RE.bed -c > ../data/PIR_overlap/`basename ${PIR_BED} .bed.gz`_overlapRE.bed\n",
    "    done\n",
    "    \n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import feature counts as a dictionary\n",
    "\n",
    "This part of the code will import the feature counts as a dictionary that can be used to map counts to features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ILC3\n",
    "ILC3_bins_H3K27ac_dict = counts_dict(\"../data/PIR_overlap/ILC_5kb_within_newbmap_CHiCAGO_ABC_peakm_filtered_PIR_overlapH3K27ac.bed\")\n",
    "ILC3_bins_H3K4me3_dict = counts_dict(\"../data/PIR_overlap/ILC_5kb_within_newbmap_CHiCAGO_ABC_peakm_filtered_PIR_overlapH3K4me3.bed\")\n",
    "ILC3_bins_ATAC_dict = counts_dict(\"../data/PIR_overlap/ILC_5kb_within_newbmap_CHiCAGO_ABC_peakm_filtered_PIR_overlapATAC.bed\")\n",
    "ILC3_bins_RE_dict = counts_dict(\"../data/PIR_overlap/ILC_5kb_within_newbmap_CHiCAGO_ABC_peakm_filtered_PIR_overlapRE.bed\")\n",
    "\n",
    "#CD4 ABC\n",
    "CD4_ABC_filtered_H3K27ac_dict = counts_dict(\"../data/PIR_overlap/CD4_1M_50K_5kb_within_newbmap_CHiCAGO_ABC_peakm_filtered_PIR_overlapH3K27ac.bed\")\n",
    "CD4_ABC_filtered_H3K4me3_dict = counts_dict(\"../data/PIR_overlap/CD4_1M_50K_5kb_within_newbmap_CHiCAGO_ABC_peakm_filtered_PIR_overlapH3K4me3.bed\")\n",
    "CD4_ABC_filtered_PIR_ATAC_dict = counts_dict(\"../data/PIR_overlap/CD4_1M_50K_5kb_within_newbmap_CHiCAGO_ABC_peakm_filtered_PIR_overlapATAC.bed\")\n",
    "CD4_ABC_filtered_PIR_RE_dict = counts_dict(\"../data/PIR_overlap/CD4_1M_50K_5kb_within_newbmap_CHiCAGO_ABC_peakm_filtered_PIR_overlapRE.bed\")\n",
    "\n",
    "#CD4 no ABC\n",
    "CD4_filtered_H3K27ac_dict = counts_dict(\"../data/PIR_overlap/CD4_1M_50K_5kb_within_newbmap_CHiCAGO_peakm_filtered_PIR_overlapH3K27ac.bed\")\n",
    "CD4_filtered_H3K4me3_dict = counts_dict(\"../data/PIR_overlap/CD4_1M_50K_5kb_within_newbmap_CHiCAGO_peakm_filtered_PIR_overlapH3K4me3.bed\")\n",
    "CD4_filtered_ATAC_dict = counts_dict(\"../data/PIR_overlap/CD4_1M_50K_5kb_within_newbmap_CHiCAGO_peakm_filtered_PIR_overlapATAC.bed\")\n",
    "CD4_filtered_RE_dict = counts_dict(\"../data/PIR_overlap/CD4_1M_50K_5kb_within_newbmap_CHiCAGO_peakm_filtered_PIR_overlapRE.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import gene counts\n",
    "\n",
    "This section will import the gene expression data and find the mean per group. The meta data for the files used can be found in [`../data/RNA/meta/20200707_CD4_ILC3_meta.txt`](../data/RNA/meta/20200707_CD4_ILC3_meta.txt)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "GENE_COUNTS = pd.read_csv(\"../data/RNA/counts/20200706_hILC3_CD4_GeneCounts.tsv\", sep=\"\\t\", header=0)\n",
    "\n",
    "GENE_COUNTS[\"ILC3_mean\"] = GENE_COUNTS[[\"SRX5797708\", \"SRX5797709\", \"SRX5797712\", \"SRX5797713\", \"SRX5797716\", \"SRX5797717\"]].mean(axis=1)\n",
    "\n",
    "GENE_COUNTS[\"CD4_mean\"] = GENE_COUNTS[[\"SRR4290846\", \"SRR4290852\", \"SRR4290853\", \"SRR4290854\"]].mean(axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map counts to interactions\n",
    "\n",
    "This section will map the number of features that overlap the PIRs. A column is created that is specific for each feature. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ILC\n",
    "ilc3.df[\"H3K27ac_counts\"] = ilc3.df[\"oe_interval_ID\"].map(ILC3_bins_H3K27ac_dict)\n",
    "ilc3.df[\"H3K4me3_counts\"] = ilc3.df[\"oe_interval_ID\"].map(ILC3_bins_H3K4me3_dict)\n",
    "ilc3.df[\"ATAC_counts\"] = ilc3.df[\"oe_interval_ID\"].map(ILC3_bins_ATAC_dict)\n",
    "ilc3.df[\"RE_counts\"] = ilc3.df[\"oe_interval_ID\"].map(ILC3_bins_RE_dict)\n",
    "\n",
    "# CD4\n",
    "cd4.df[\"H3K27ac_counts\"] = cd4.df[\"oe_interval_ID\"].map(CD4_filtered_H3K27ac_dict)\n",
    "cd4.df[\"H3K4me3_counts\"] = cd4.df[\"oe_interval_ID\"].map(CD4_filtered_H3K4me3_dict)\n",
    "cd4.df[\"ATAC_counts\"] = cd4.df[\"oe_interval_ID\"].map(CD4_filtered_ATAC_dict)\n",
    "cd4.df[\"RE_counts\"] = cd4.df[\"oe_interval_ID\"].map(CD4_filtered_RE_dict)\n",
    "\n",
    "# CD4 ABC\n",
    "cd4_abc.df[\"H3K27ac_counts\"] = cd4_abc.df[\"oe_interval_ID\"].map(CD4_ABC_filtered_H3K27ac_dict)\n",
    "cd4_abc.df[\"H3K4me3_counts\"] = cd4_abc.df[\"oe_interval_ID\"].map(CD4_ABC_filtered_H3K4me3_dict)\n",
    "cd4_abc.df[\"ATAC_counts\"] = cd4_abc.df[\"oe_interval_ID\"].map(CD4_ABC_filtered_PIR_ATAC_dict)\n",
    "cd4_abc.df[\"RE_counts\"] = cd4_abc.df[\"oe_interval_ID\"].map(CD4_ABC_filtered_PIR_RE_dict)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map Feature Counts to Genes\n",
    "\n",
    "We will then map the different feature counts to the genes that we are interested in. We will create new columns in the gene expression dataframe that contain this information. \n",
    "\n",
    "* Enhancer counts are the number of PIRs that overlap the Gene. This uses the count function. \n",
    "* The H3K27ac, H3K4me3, ATAC, and RE features are summed up across PIRs. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "GENE_COUNTS[\"ILC3_enhancer_count_bins\"] = GENE_COUNTS[\"GeneName\"].map(ilc3.df.groupby([\"baitName\"]).count()[\"baitChr\"])\n",
    "GENE_COUNTS[\"ILC3_H3K27ac_count_bins\"] = GENE_COUNTS[\"GeneName\"].map(ilc3.df.groupby([\"baitName\"]).sum()[\"H3K27ac_counts\"])\n",
    "GENE_COUNTS[\"ILC3_H3K4me3_count_bins\"] = GENE_COUNTS[\"GeneName\"].map(ilc3.df.groupby([\"baitName\"]).sum()[\"H3K4me3_counts\"])\n",
    "GENE_COUNTS[\"ILC3_ATAC_count_bins\"] = GENE_COUNTS[\"GeneName\"].map(ilc3.df.groupby([\"baitName\"]).sum()[\"ATAC_counts\"])\n",
    "GENE_COUNTS[\"ILC3_RE_count_bins\"] = GENE_COUNTS[\"GeneName\"].map(ilc3.df.groupby([\"baitName\"]).sum()[\"RE_counts\"])\n",
    "\n",
    "GENE_COUNTS[\"CD4_enhancer_counts_bins\"] = GENE_COUNTS[\"GeneName\"].map(cd4.df.groupby([\"baitName\"]).count()[\"baitChr\"])\n",
    "GENE_COUNTS[\"CD4_H3K27ac_count_bins\"] = GENE_COUNTS[\"GeneName\"].map(cd4.df.groupby([\"baitName\"]).sum()[\"H3K27ac_counts\"])\n",
    "GENE_COUNTS[\"CD4_H3K4me3_count_bins\"] = GENE_COUNTS[\"GeneName\"].map(cd4.df.groupby([\"baitName\"]).sum()[\"H3K4me3_counts\"])\n",
    "GENE_COUNTS[\"CD4_ATAC_count_bins\"] = GENE_COUNTS[\"GeneName\"].map(cd4.df.groupby([\"baitName\"]).sum()[\"ATAC_counts\"])\n",
    "GENE_COUNTS[\"CD4_RE_count_bins\"] = GENE_COUNTS[\"GeneName\"].map(cd4.df.groupby([\"baitName\"]).sum()[\"RE_counts\"])\n",
    "\n",
    "GENE_COUNTS[\"CD4_ABC_enhancer_count_bins\"] = GENE_COUNTS[\"GeneName\"].map(cd4_abc.df.groupby([\"baitName\"]).count()[\"baitChr\"])\n",
    "GENE_COUNTS[\"CD4_ABC_H3K27ac_count_bins\"] = GENE_COUNTS[\"GeneName\"].map(cd4_abc.df.groupby([\"baitName\"]).sum()[\"H3K27ac_counts\"])\n",
    "GENE_COUNTS[\"CD4_ABC_H3K4me3_count_bins\"] = GENE_COUNTS[\"GeneName\"].map(cd4_abc.df.groupby([\"baitName\"]).sum()[\"H3K4me3_counts\"])\n",
    "GENE_COUNTS[\"CD4_ABC_ATAC_count_bins\"] = GENE_COUNTS[\"GeneName\"].map(cd4_abc.df.groupby([\"baitName\"]).sum()[\"ATAC_counts\"])\n",
    "GENE_COUNTS[\"CD4_ABC_RE_count_bins\"] = GENE_COUNTS[\"GeneName\"].map(cd4_abc.df.groupby([\"baitName\"]).sum()[\"RE_counts\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneName</th>\n",
       "      <th>SRX5797708</th>\n",
       "      <th>SRX5797709</th>\n",
       "      <th>SRX5797712</th>\n",
       "      <th>SRX5797713</th>\n",
       "      <th>SRX5797716</th>\n",
       "      <th>SRX5797717</th>\n",
       "      <th>SRR4290846</th>\n",
       "      <th>SRR4290852</th>\n",
       "      <th>SRR4290853</th>\n",
       "      <th>...</th>\n",
       "      <th>CD4_enhancer_counts_bins</th>\n",
       "      <th>CD4_H3K27ac_count_bins</th>\n",
       "      <th>CD4_H3K4me3_count_bins</th>\n",
       "      <th>CD4_ATAC_count_bins</th>\n",
       "      <th>CD4_RE_count_bins</th>\n",
       "      <th>CD4_ABC_enhancer_count_bins</th>\n",
       "      <th>CD4_ABC_H3K27ac_count_bins</th>\n",
       "      <th>CD4_ABC_H3K4me3_count_bins</th>\n",
       "      <th>CD4_ABC_ATAC_count_bins</th>\n",
       "      <th>CD4_ABC_RE_count_bins</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>876</td>\n",
       "      <td>508</td>\n",
       "      <td>575</td>\n",
       "      <td>276</td>\n",
       "      <td>222</td>\n",
       "      <td>387</td>\n",
       "      <td>774</td>\n",
       "      <td>777</td>\n",
       "      <td>474</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>36</td>\n",
       "      <td>31</td>\n",
       "      <td>38</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M</td>\n",
       "      <td>20</td>\n",
       "      <td>50</td>\n",
       "      <td>44</td>\n",
       "      <td>66</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>1122</td>\n",
       "      <td>638</td>\n",
       "      <td>172</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2ML1</td>\n",
       "      <td>37</td>\n",
       "      <td>84</td>\n",
       "      <td>76</td>\n",
       "      <td>102</td>\n",
       "      <td>49</td>\n",
       "      <td>65</td>\n",
       "      <td>80</td>\n",
       "      <td>90</td>\n",
       "      <td>78</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A2MP1</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>273</td>\n",
       "      <td>168</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22454</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>232</td>\n",
       "      <td>354</td>\n",
       "      <td>129</td>\n",
       "      <td>156</td>\n",
       "      <td>485</td>\n",
       "      <td>264</td>\n",
       "      <td>885</td>\n",
       "      <td>772</td>\n",
       "      <td>622</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22455</th>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>71</td>\n",
       "      <td>179</td>\n",
       "      <td>89</td>\n",
       "      <td>141</td>\n",
       "      <td>76</td>\n",
       "      <td>148</td>\n",
       "      <td>306</td>\n",
       "      <td>275</td>\n",
       "      <td>232</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22456</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>65</td>\n",
       "      <td>908</td>\n",
       "      <td>577</td>\n",
       "      <td>719</td>\n",
       "      <td>84</td>\n",
       "      <td>440</td>\n",
       "      <td>2231</td>\n",
       "      <td>1996</td>\n",
       "      <td>1426</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22457</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>513</td>\n",
       "      <td>703</td>\n",
       "      <td>180</td>\n",
       "      <td>452</td>\n",
       "      <td>196</td>\n",
       "      <td>501</td>\n",
       "      <td>1645</td>\n",
       "      <td>1788</td>\n",
       "      <td>1678</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22458</th>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>186</td>\n",
       "      <td>1074</td>\n",
       "      <td>308</td>\n",
       "      <td>596</td>\n",
       "      <td>114</td>\n",
       "      <td>706</td>\n",
       "      <td>3257</td>\n",
       "      <td>2963</td>\n",
       "      <td>2773</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22459 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      GeneName  SRX5797708  SRX5797709  SRX5797712  SRX5797713  SRX5797716  \\\n",
       "0         A1BG         876         508         575         276         222   \n",
       "1         A1CF           1           8           5           6           4   \n",
       "2          A2M          20          50          44          66           0   \n",
       "3        A2ML1          37          84          76         102          49   \n",
       "4        A2MP1           0          21           1           4           0   \n",
       "...        ...         ...         ...         ...         ...         ...   \n",
       "22454     ZXDC         232         354         129         156         485   \n",
       "22455   ZYG11A          71         179          89         141          76   \n",
       "22456   ZYG11B          65         908         577         719          84   \n",
       "22457      ZYX         513         703         180         452         196   \n",
       "22458    ZZEF1         186        1074         308         596         114   \n",
       "\n",
       "       SRX5797717  SRR4290846  SRR4290852  SRR4290853  ...  \\\n",
       "0             387         774         777         474  ...   \n",
       "1               7          36          31          38  ...   \n",
       "2              21        1122         638         172  ...   \n",
       "3              65          80          90          78  ...   \n",
       "4               3         273         168          11  ...   \n",
       "...           ...         ...         ...         ...  ...   \n",
       "22454         264         885         772         622  ...   \n",
       "22455         148         306         275         232  ...   \n",
       "22456         440        2231        1996        1426  ...   \n",
       "22457         501        1645        1788        1678  ...   \n",
       "22458         706        3257        2963        2773  ...   \n",
       "\n",
       "       CD4_enhancer_counts_bins  CD4_H3K27ac_count_bins  \\\n",
       "0                           NaN                     NaN   \n",
       "1                           3.0                     0.0   \n",
       "2                           NaN                     NaN   \n",
       "3                           2.0                     0.0   \n",
       "4                           NaN                     NaN   \n",
       "...                         ...                     ...   \n",
       "22454                       NaN                     NaN   \n",
       "22455                       NaN                     NaN   \n",
       "22456                       2.0                     0.0   \n",
       "22457                       NaN                     NaN   \n",
       "22458                       NaN                     NaN   \n",
       "\n",
       "       CD4_H3K4me3_count_bins  CD4_ATAC_count_bins  CD4_RE_count_bins  \\\n",
       "0                         NaN                  NaN                NaN   \n",
       "1                         0.0                  0.0                0.0   \n",
       "2                         NaN                  NaN                NaN   \n",
       "3                         0.0                  0.0                0.0   \n",
       "4                         NaN                  NaN                NaN   \n",
       "...                       ...                  ...                ...   \n",
       "22454                     NaN                  NaN                NaN   \n",
       "22455                     NaN                  NaN                NaN   \n",
       "22456                     1.0                  2.0                2.0   \n",
       "22457                     NaN                  NaN                NaN   \n",
       "22458                     NaN                  NaN                NaN   \n",
       "\n",
       "       CD4_ABC_enhancer_count_bins  CD4_ABC_H3K27ac_count_bins  \\\n",
       "0                              2.0                         0.0   \n",
       "1                              5.0                         0.0   \n",
       "2                              2.0                         1.0   \n",
       "3                              3.0                         0.0   \n",
       "4                              NaN                         NaN   \n",
       "...                            ...                         ...   \n",
       "22454                          NaN                         NaN   \n",
       "22455                          2.0                         1.0   \n",
       "22456                          5.0                         2.0   \n",
       "22457                          6.0                         0.0   \n",
       "22458                          NaN                         NaN   \n",
       "\n",
       "       CD4_ABC_H3K4me3_count_bins  CD4_ABC_ATAC_count_bins  \\\n",
       "0                             1.0                      3.0   \n",
       "1                             2.0                      6.0   \n",
       "2                             0.0                      2.0   \n",
       "3                             1.0                      4.0   \n",
       "4                             NaN                      NaN   \n",
       "...                           ...                      ...   \n",
       "22454                         NaN                      NaN   \n",
       "22455                         1.0                      7.0   \n",
       "22456                         2.0                     13.0   \n",
       "22457                         0.0                     10.0   \n",
       "22458                         NaN                      NaN   \n",
       "\n",
       "       CD4_ABC_RE_count_bins  \n",
       "0                        3.0  \n",
       "1                        6.0  \n",
       "2                        2.0  \n",
       "3                        4.0  \n",
       "4                        NaN  \n",
       "...                      ...  \n",
       "22454                    NaN  \n",
       "22455                    6.0  \n",
       "22456                   11.0  \n",
       "22457                   10.0  \n",
       "22458                    NaN  \n",
       "\n",
       "[22459 rows x 28 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GENE_COUNTS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot Gene Expression Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### ILC3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "ILC3_expression_atac = GeneExpressionAnalysis(GENE_COUNTS, \n",
    "                                                               feature_col=\"ILC3_ATAC_count_bins\", \n",
    "                                                               mean_col=\"ILC3_mean\",\n",
    "                                                               output_filename=\"../data/expression_PIR_analysis/ILC3_mean_expression_PIRwATAC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "ILC3_expression_acety = GeneExpressionAnalysis(GENE_COUNTS, \n",
    "                                                               feature_col=\"ILC3_H3K27ac_count_bins\", \n",
    "                                                               mean_col=\"ILC3_mean\",\n",
    "                                                               output_filename=\"../data/expression_PIR_analysis/ILC3_mean_expression_PIRwH3K27ac\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "ILC3_expression_methyl = GeneExpressionAnalysis(GENE_COUNTS, \n",
    "                                                               feature_col=\"ILC3_H3K4me3_count_bins\", \n",
    "                                                               mean_col=\"ILC3_mean\",\n",
    "                                                               output_filename=\"../data/expression_PIR_analysis/ILC3_mean_expression_PIRwH3K4me3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "ILC3_expression_enhancer = GeneExpressionAnalysis(GENE_COUNTS, \n",
    "                                                               feature_col=\"ILC3_ATAC_count_bins\", \n",
    "                                                               mean_col=\"ILC3_mean\",\n",
    "                                                               output_filename=\"../data/expression_PIR_analysis/ILC3_mean_expression_PIRwEnhancer\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### CD4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd4_expression_atac = GeneExpressionAnalysis(GENE_COUNTS, \n",
    "                                                               feature_col=\"CD4_ATAC_count_bins\", \n",
    "                                                               mean_col=\"CD4_mean\",\n",
    "                                                               output_filename=\"../data/expression_PIR_analysis/CD4_mean_expression_PIRwATAC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd4_expression_acety = GeneExpressionAnalysis(GENE_COUNTS, \n",
    "                                                               feature_col=\"CD4_H3K27ac_count_bins\", \n",
    "                                                               mean_col=\"CD4_mean\",\n",
    "                                                               output_filename=\"../data/expression_PIR_analysis/CD4_mean_expression_PIRwH3K27ac\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd4_expression_methyl = GeneExpressionAnalysis(GENE_COUNTS, \n",
    "                                                               feature_col=\"CD4_H3K4me3_count_bins\", \n",
    "                                                               mean_col=\"CD4_mean\",\n",
    "                                                               output_filename=\"../data/expression_PIR_analysis/CD4_mean_expression_PIRwH3K4me3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd4_expression_enhancer = GeneExpressionAnalysis(GENE_COUNTS, \n",
    "                                                               feature_col=\"CD4_enhancer_counts_bins\", \n",
    "                                                               mean_col=\"CD4_mean\",\n",
    "                                                               output_filename=\"../data/expression_PIR_analysis/CD4_mean_expression_PIRwEnhancer\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### CD4 ABC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd4_abc_expression_atac = GeneExpressionAnalysis(GENE_COUNTS, \n",
    "                                                               feature_col=\"CD4_ABC_ATAC_count_bins\", \n",
    "                                                               mean_col=\"CD4_mean\",\n",
    "                                                               output_filename=\"../data/expression_PIR_analysis/CD4_ABC_mean_expression_PIRwATAC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd4_abc_expression_acety = GeneExpressionAnalysis(GENE_COUNTS, \n",
    "                                                               feature_col=\"CD4_ABC_H3K27ac_count_bins\", \n",
    "                                                               mean_col=\"CD4_mean\",\n",
    "                                                               output_filename=\"../data/expression_PIR_analysis/CD4_ABC_mean_expression_PIRwH3K27ac\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd4_abc_expression_methyl = GeneExpressionAnalysis(GENE_COUNTS, \n",
    "                                                               feature_col=\"CD4_ABC_H3K4me3_count_bins\", \n",
    "                                                               mean_col=\"CD4_mean\",\n",
    "                                                               output_filename=\"../data/expression_PIR_analysis/CD4_ABC_mean_expression_PIRwH3K4me3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd4_abc_expression_enhancer = GeneExpressionAnalysis(GENE_COUNTS, \n",
    "                                                               feature_col=\"CD4_ABC_enhancer_count_bins\", \n",
    "                                                               mean_col=\"CD4_mean\",\n",
    "                                                               output_filename=\"../data/expression_PIR_analysis/CD4_ABC_mean_expression_PIRwEnhancer\")"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "c40e1b489fb11c0bd8cc5a783b5d3b77609a7988a043d25bde579eefe634d11a"
  },
  "kernelspec": {
   "display_name": "Python 3.9.10 ('maxatac')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
