{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of pcHi-C data from human CD4+ T cells and ILC3 cells"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook will walk through the analysis of pcHi-C data generated by Mikhail Spivakov's group. We will also analyze CD4+ alpha-beta T-cells in addition to the ILC3 data as a positive control sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Overview\n",
    "We have two cell types to analyze:\n",
    "\n",
    "* Human Tonsil ILC3\n",
    "* Human Venous Blood CD4+ alpha/beta T-cells\n",
    "\n",
    "These two cell types were analyzed with two different analysis methods:\n",
    "\n",
    "* 5kb bin level - interactions are approximately 5k bins instead of the size of restriction fragments\n",
    "* fragment level - interactions are the size of the restriction fragment made by DPNII\n",
    "\n",
    "### Update 06.06.2022\n",
    "Valeriya emailed me [new data](https://www.transferxl.com/manage/089FGscZhgKG8?token=6ao3nPY77S1Ed5_eBlxd3nYkCu-fgYAUBqjUtj9s74yAogdy2o8_QW0W1gq6wcdI-A%2C%2C) to analyze using the following approach:\n",
    " \n",
    "> You will see that I also attached a CD4 CHiCAGO peakmatrix without ABC, because I now rerun CHiCAGO with more biological replicates merged together. If you could re-run RELI for that file too, it would be really great. Thank you!!\n",
    "\n",
    "> Removing trans-chromosomal interactions can be achieved by removing the NA distance values. There should be no interactinos with dist = 0 in the Chicago peakmatrices. The Capture Hi-C-adapted ABC analysis can generate pairs with dist = 0 and I should have removed those already from the peakmatrices that I sent you, but it wouldn’t harm to remove again in your pipeline just in case.\n",
    "\n",
    ">CHiCAGO score >=5 is oerfect, when working with peakmatrices generated from CHiCAGO interactions only. However, for the pekamatrices, containing both CHiCAGO contacts and ABC pairs, please use the merged_score column >=5 for filtering.\n",
    "\n",
    ">For the CD4+ data I have now run CHiCAGO on 4 replicates (two reps for 1M and two reps for 50K). So, for filtering you should just use either the score scolumn (in case of the CHiCAGO peakmatrix) and merged_score column (in case of the CHiCAGO_ABC peakmatrix). So there is no need to look at the scores in previous peakmatrices.\n",
    "\n",
    ">please ignore the remove_line column.\n",
    "\n",
    ">don’t worry about the off-targets in the OEName column.\n",
    "\n",
    ">The promoter-promoter interactions should be removed for the RELI analysis. Have you noticed any difference for when you run with and without these interactions?\n",
    "___\n",
    "\n",
    "## Data\n",
    "\n",
    "### RNA-seq\n",
    "\n",
    "* ILC3 RNA-seq data was gathered from [GSE130775](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE130775). \n",
    "* CD4+ gene expression was downloaded from [GSE87254](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE87254).\n",
    "\n",
    "#### Approach\n",
    "1. I used Salmon to quasi-map reads to transcripts. The data was aligned to hg38 annotations from ENSEMBL. I used the following site to fetch the [hg38 genome](http://uswest.ensembl.org/Homo_sapiens/Info/Index). I used the site [HBC Github Salmon](https://hbctraining.github.io/Intro-to-rnaseq-hpc-salmon/lessons/04_quasi_alignment_salmon.html) tutorial as a guide for my analysis.\n",
    "\n",
    "2. I imported the transcript counts and collapsed them to gene counts using [Tx import](https://bioconductor.statistik.tu-dortmund.de/packages/33/bioc/vignettes/tximport/inst/doc/tximport.html) as described in the vignette.\n",
    "\n",
    "3. I exported the gene expression matrix for ILC3 here: `20200706_hILC3_CD4_GeneCounts.tsv`.\n",
    "\n",
    "### ChIP-seq\n",
    "\n",
    "* I obtained ILC3 histone marker data from [GSE77299](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi) \n",
    "* CD4+ histone marker data from [Blueprint Project](http://dcc.blueprint-epigenome.eu/#/experiments). \n",
    "\n",
    "Detailed methods can be found in the write-up.\n",
    "\n",
    "### ATAC-seq\n",
    "\n",
    "1) I obtained ILC3 ATAC-seq data from [GSE77299](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi) and CD4+ ATAC-seq data was gathered from [PRJNA380283](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5623106/). \n",
    "2) The CD4+ ATAC-seq data was mapped to hg19 so UCSC `liftover` was used to convert the ATAC-seq data to hg38. Detailed methods can be found in the write-up.\n",
    "\n",
    "### Regulatory Elements\n",
    "\n",
    "The RE features used in this analysis are the union of ATAC-seq data with H3K27ac data.\n",
    "\n",
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting PIRs From CHiCAGO output file\n",
    "\n",
    "I wrote a python class object to work with the CHiCAGO output text file. This piece of code will perform filtering of specific types of interactions, like promoter-to-promoter, or trans-chromosomal interactions. \n",
    "\n",
    "Example:\n",
    "\n",
    "```python\n",
    "input_file = \"/Users/caz3so/scratch/20220606_spivakov_pchic_reanalysis/TransferXL-089FGscZhgKG8/ILC_5kb_within_newbmap_CHiCAGO_ABC_peakm.txt\"\n",
    "\n",
    "ILC3_data = ChicagoData(input_file)\n",
    "\n",
    "ILC3_data.df(input_file, dropna=True, drop_off_target=True, drop_p2p=False, drop_trans_chrom=True)\n",
    "\n",
    "ILC3_data.pir_df\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ChicagoData(object):\n",
    "    \"\"\"Import CHICACO data\n",
    "    \"\"\"\n",
    "    def __init__(self,\n",
    "                 filename: str,\n",
    "                 dropna: bool = True,\n",
    "                 drop_off_target: bool = True,\n",
    "                 drop_trans_chrom: bool = True,\n",
    "                 score_col: str = None\n",
    "                 ):\n",
    "        \"\"\"Initialize the object\n",
    "\n",
    "        Args:\n",
    "            filename (str): Input CHICAGO txt file\n",
    "            dropna (bool, optional): Drop the interactions with NA. Defaults to =True.\n",
    "            drop_off_target (bool, optional): Drop off target interactions. Defaults to =True.\n",
    "            drop_trans_chrom (bool, optional): Drop transchromosomal interactions. Defaults to =True.\n",
    "        \"\"\"\n",
    "        # Set filename to the input filename\n",
    "        self.filename = filename\n",
    "        # Set whether to drop na interactions\n",
    "        self.dropna = dropna\n",
    "        # Set whether to drop off target\n",
    "        self.drop_off_target = drop_off_target\n",
    "        # Set whether to drop transchromosomal interactions\n",
    "        self.drop_trans_chrom = drop_trans_chrom\n",
    "        # Score column name for filtering\n",
    "        self.score_col = score_col\n",
    "        \n",
    "        # Read file into DF\n",
    "        self._read_file_()\n",
    "        \n",
    "        # Format the DF\n",
    "        self._format_file_()\n",
    "        \n",
    "        # Filter the formatted DF\n",
    "        self._filter_file_()\n",
    "\n",
    "        # Get the PIR df\n",
    "        self._get_PIR_df_()\n",
    "        \n",
    "    def _read_file_(self):\n",
    "        \"\"\"Read in original file\n",
    "        \"\"\"\n",
    "        # Read in original file and save\n",
    "        self.input_df =  pd.read_csv(self.filename, sep=\"\\t\", header=0, low_memory=False)\n",
    "    \n",
    "    def _format_file_(self):\n",
    "        \"\"\"Format CHICAGO file\n",
    "        \"\"\"\n",
    "        # Create a copy of the raw input to be manipulated\n",
    "        df = self.input_df.copy()\n",
    "        \n",
    "        # Format the chromosome names\n",
    "        df[\"baitChr\"] = \"chr\" + df[\"baitChr\"].apply(str)\n",
    "        df[\"oeChr\"] = \"chr\" + df[\"oeChr\"].apply(str)\n",
    "        \n",
    "        # Calculate the wide of the intervals\n",
    "        df[\"OE_width\"] = df[\"oeEnd\"] - df[\"oeStart\"]\n",
    "        \n",
    "        # Create an ID column that can be used to track the intervals\n",
    "        df[\"ID\"] = df[\"baitChr\"] + \":\" + \\\n",
    "                   df[\"baitStart\"].apply(str) + \"-\" + \\\n",
    "                   df[\"baitEnd\"].apply(str) + \"_\" + \\\n",
    "                   df[\"oeChr\"] + \":\" + \\\n",
    "                   df[\"oeStart\"].apply(str) + \"-\" + \\\n",
    "                   df[\"oeEnd\"].apply(str)\n",
    "\n",
    "        # Set the variable to the formatted df\n",
    "        self.df = df\n",
    "    \n",
    "    def _filter_file_(self):\n",
    "        \"\"\"Filter the formatted CHICAGO results\n",
    "        \"\"\"\n",
    "        if self.dropna:\n",
    "            self.df.dropna(subset=[\"dist\"], inplace=True)\n",
    "\n",
    "        if self.drop_off_target:\n",
    "            self.df[self.df[\"baitName\"] != \"off_target\"]\n",
    "\n",
    "        if self.drop_trans_chrom:\n",
    "            self.df = self.df[self.df[\"baitChr\"] == self.df[\"oeChr\"]]\n",
    "        \n",
    "        if self.score_col:\n",
    "            self.df = self.df[self.df[self.score_col] >= 5]\n",
    "\n",
    "        \n",
    "    def _get_PIR_df_(self):\n",
    "        \"\"\"Get a DF of all PIR interactions\n",
    "        \"\"\"\n",
    "        self.pir_df = self.df[[\"oeChr\", \"oeStart\", \"oeEnd\", \"OE_width\"]].drop_duplicates(subset=[\"oeChr\", \"oeStart\", \"oeEnd\"], keep=\"first\")\n",
    "\n",
    "    def write_PIR_bed(self, output_filename): \n",
    "        \"\"\"Write PIRs from the filtered CHICAGO results to a bed file\n",
    "        \"\"\"       \n",
    "        self.pir_df.to_csv(output_filename, sep=\"\\t\", header=False, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "c40e1b489fb11c0bd8cc5a783b5d3b77609a7988a043d25bde579eefe634d11a"
  },
  "kernelspec": {
   "display_name": "Python 3.9.10 ('maxatac')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
