{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of pcHi-C data from human CD4+ T cells and ILC3 cells\n",
    "\n",
    "This notebook will walk through the analysis of pcHi-C data generated by Mikhail Spivakov's group. We will also analyze CD4+ alpha-beta T-cells in addition to the ILC3 data as a positive control sample."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "from sklearn.metrics import r2_score\n",
    "from scipy.stats import pearsonr\n",
    "from scipy.stats import spearmanr\n",
    "from decimal import Decimal\n",
    "import matplotlib.ticker as ticker\n",
    "import pybedtools\n",
    "import os\n",
    "from os import makedirs, error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ChicagoData(object):\n",
    "    \"\"\"Import CHiCAGO data\n",
    "    \"\"\"\n",
    "    def __init__(self,\n",
    "                 filename: str,\n",
    "                 drop_off_target_bait: bool = True,\n",
    "                 drop_off_target_oe: bool = True,\n",
    "                 drop_trans_chrom: bool = True,\n",
    "                 score_col: str = None,\n",
    "                 score_val: int = 5,\n",
    "                 remove_p2p: bool= True,\n",
    "                 features_to_count: dict = {},\n",
    "                 gene_expression: str = \"\",\n",
    "                 nonzero_expression: bool = True,\n",
    "                 dropna_expression: bool = True,\n",
    "                 output_dir: str = \"\",\n",
    "                 output_basename: str = \"\"\n",
    "                 ):\n",
    "        \"\"\"Initialize the object\n",
    "\n",
    "        Args:\n",
    "            filename (str): Input CHICAGO txt file\n",
    "            drop_off_target (bool, optional): Drop off target interactions. Defaults to =True.\n",
    "            drop_trans_chrom (bool, optional): Drop transchromosomal interactions. Defaults to =True.\n",
    "        \"\"\"\n",
    "        # Set filename to the input filename\n",
    "        self.filename = filename\n",
    "        # Set whether to drop off target baits\n",
    "        self.drop_off_target_bait = drop_off_target_bait\n",
    "        # Set whether to drop off target oe\n",
    "        self.drop_off_target_oe = drop_off_target_oe\n",
    "        # Set whether to drop transchromosomal interactions\n",
    "        self.drop_trans_chrom = drop_trans_chrom\n",
    "        # Score column name for filtering\n",
    "        self.score_col = score_col\n",
    "        # Score value threshold\n",
    "        self.score_val = score_val\n",
    "        # Remove promoter to promoter interactions\n",
    "        self.remove_p2p = remove_p2p\n",
    "        # Map feature counts to PIR if provided\n",
    "        self.features_to_count = features_to_count\n",
    "        # Import the gene expression matrix\n",
    "        self.gene_expression = gene_expression\n",
    "        # Only keep non-zero expression\n",
    "        self.nonzero_expression = nonzero_expression\n",
    "        # Drop na Values\n",
    "        self.dropna_expression = dropna_expression\n",
    "        # Set the output directory\n",
    "        self.output_dir = output_dir\n",
    "        # Set the output basename \n",
    "        self.basename = output_basename\n",
    "        \n",
    "        # Read file into DF\n",
    "        self._read_file_()\n",
    "        \n",
    "        # Format the DF\n",
    "        self._format_file_()\n",
    "        \n",
    "        # Filter the formatted DF\n",
    "        self._filter_file_()\n",
    "        \n",
    "        # Get the PIR df\n",
    "        self._get_PIR_df_()\n",
    "        \n",
    "        # Get the bait df\n",
    "        self._get_bait_df_()\n",
    "        \n",
    "        # Get the combined df\n",
    "        self._get_combined_df_()  \n",
    "        \n",
    "        self._get_feature_counts_()\n",
    "        \n",
    "        self._import_gene_counts_()\n",
    "            \n",
    "        self._map_feature_counts_to_genes_()\n",
    "        \n",
    "        self._filter_expression_()\n",
    "        \n",
    "        self._get_PIR_count_v_mean_()\n",
    "        \n",
    "        self._write_new_chicago_data_()\n",
    "                    \n",
    "    def _read_file_(self):\n",
    "        \"\"\"Read in original file\n",
    "        \"\"\"\n",
    "        # Read in original file and save\n",
    "        self.input_df =  pd.read_csv(self.filename, sep=\"\\t\", header=0, low_memory=False)\n",
    "    \n",
    "    def _write_new_chicago_data_(self):\n",
    "        output_dir = os.path.join(self.output_dir, \"modified_chicago\")\n",
    "        \n",
    "        output_filename = os.path.join(output_dir, f\"{self.basename}_modified.tsv\")\n",
    "        \n",
    "        self._get_dir_(output_dir)\n",
    "\n",
    "        self.df.to_csv(output_filename, sep=\"\\t\", index=False)\n",
    "        \n",
    "    def _format_file_(self):\n",
    "        \"\"\"Format CHICAGO file\n",
    "        \"\"\"\n",
    "        # Create a copy of the raw input to be manipulated\n",
    "        df = self.input_df.copy()\n",
    "        \n",
    "        # Format the chromosome names\n",
    "        df[\"baitChr\"] = \"chr\" + df[\"baitChr\"].apply(str)\n",
    "        df[\"oeChr\"] = \"chr\" + df[\"oeChr\"].apply(str)\n",
    "        \n",
    "        df[\"oe_interval_ID\"] = df[\"oeChr\"] + \":\" + \\\n",
    "                   df[\"oeStart\"].apply(str) + \"-\" + \\\n",
    "                   df[\"oeEnd\"].apply(str)\n",
    "\n",
    "        df[\"bait_interval_ID\"] = df[\"baitChr\"] + \":\" + \\\n",
    "                   df[\"baitStart\"].apply(str) + \"-\" + \\\n",
    "                   df[\"baitEnd\"].apply(str)\n",
    "\n",
    "        df[\"interaction_ID\"] = df[\"bait_interval_ID\"] + \"_\" + df[\"oe_interval_ID\"].apply(str)\n",
    "        \n",
    "        # Set the variable to the formatted df\n",
    "        self.bait_ID = df[\"baitID\"].unique()\n",
    "        \n",
    "        self.oe_ID = df[\"oeID\"].unique()\n",
    "\n",
    "        self.bait_interval_ID = df[\"bait_interval_ID\"].unique()\n",
    "        \n",
    "        self.oe_interval_ID = df[\"oe_interval_ID\"].unique()\n",
    "\n",
    "        self.df = df\n",
    "\n",
    "    def _filter_file_(self):\n",
    "        \"\"\"Filter the formatted CHICAGO results\n",
    "        \"\"\"\n",
    "        # Drop the off target baits\n",
    "        if self.drop_off_target_bait:\n",
    "            self.df[self.df[\"baitName\"] != \"off_target\"]\n",
    "\n",
    "        # Drop the off target OE names\n",
    "        if self.drop_off_target_oe:\n",
    "            self.df[self.df[\"oeName\"] != \"off_target\"]\n",
    "\n",
    "        # Drop the trans chromosomal interactions\n",
    "        if self.drop_trans_chrom:\n",
    "            self.df = self.df[self.df[\"dist\"] != \".\"]\n",
    "                        \n",
    "            self.df = self.df[self.df[\"dist\"].apply(float) != 0]\n",
    "\n",
    "            self.df = self.df.dropna(subset=[\"dist\"])\n",
    "\n",
    "            self.df = self.df[self.df[\"baitChr\"] == self.df[\"oeChr\"]]\n",
    "\n",
    "        # Filter the specific score column by a specific value\n",
    "        if self.score_col:\n",
    "            self.df = self.df[self.df[self.score_col] >= self.score_val]\n",
    "            \n",
    "        # Drop promoter to promoter interactions\n",
    "        if self.remove_p2p:\n",
    "            self.df = self.df[self.df.oeName == \".\"]\n",
    "            \n",
    "            self.df = self.df[~self.df.oe_interval_ID.isin(self.bait_interval_ID)]\n",
    "        \n",
    "    def _get_PIR_df_(self):\n",
    "        \"\"\"Get a DF of all PIR interactions\n",
    "        \"\"\"\n",
    "        self.pir_df = self.df[[\"oeChr\", \"oeStart\", \"oeEnd\", \"oe_interval_ID\"]].drop_duplicates(subset=[\"oeChr\", \"oeStart\", \"oeEnd\"], keep=\"first\")\n",
    "        \n",
    "        self.PIR_bt = pybedtools.BedTool.from_dataframe(self.pir_df)\n",
    "\n",
    "    def _get_bait_df_(self):\n",
    "        \"\"\"Get a DF of baits\n",
    "        \"\"\"\n",
    "        self.bait_df = self.df[[\"baitChr\", \"baitStart\", \"baitEnd\", \"bait_interval_ID\"]].drop_duplicates(subset=[\"baitChr\", \"baitStart\", \"baitEnd\"], keep=\"first\")\n",
    "        \n",
    "    def _get_combined_df_(self):\n",
    "        \"\"\"Get a comined DF\n",
    "        \"\"\"\n",
    "        tmp_df = self.pir_df.copy()\n",
    "        tmp_df2 = self.bait_df.copy()\n",
    "\n",
    "        tmp_df.columns = [\"Chr\", \"Start\", \"Stop\", \"ID\"]\n",
    "        tmp_df2.columns = [\"Chr\", \"Start\", \"Stop\", \"ID\"]\n",
    "\n",
    "        self.unique_features = pd.concat([tmp_df, tmp_df2])\n",
    "    \n",
    "    def _get_feature_counts_(self):\n",
    "        \"\"\"Get the counts of features that overlap PIRs and map them back to the pcHiC interaction\n",
    "        \"\"\"\n",
    "        for file, tag in self.features_to_count.items():\n",
    "            print(f\"Importing {file} : Column will be saved as {tag}\")\n",
    "            output_intersection_dir = os.path.join(self.output_dir, \"PIR_intersection\")\n",
    "\n",
    "            output_intersection_fname = os.path.join(output_intersection_dir, f\"{self.basename}_PIR_intersect_{tag}.bed\")\n",
    "            \n",
    "            # Import and convert the featuers to a bedtools\n",
    "            feature_bt = pybedtools.BedTool(file)\n",
    "            \n",
    "            # Intersect the features to get counts\n",
    "            feature_counts = self.PIR_bt.intersect(feature_bt, c=True)\n",
    "            \n",
    "            # Intersect the features to get overlaps (true intersections)\n",
    "            feature_intersection = self.PIR_bt.intersect(feature_bt)\n",
    "            \n",
    "            feature_intersection_sort = feature_intersection.sort()\n",
    "            \n",
    "            # Convert bedtools to pandas\n",
    "            feature_counts_df = feature_counts.to_dataframe()\n",
    "                        \n",
    "            # Convert bedtools to dataframe\n",
    "            feature_intersection_df = feature_intersection_sort.to_dataframe()\n",
    "                        \n",
    "            # Create a dictionary of counts \n",
    "            counts_dict =  pd.Series(feature_counts_df[\"score\"].values,index=feature_counts_df[\"name\"]).to_dict()\n",
    "\n",
    "            # Map the counts back to the CHICAGO dataframe\n",
    "            self.df[tag] = self.df[\"oe_interval_ID\"].map(counts_dict)\n",
    "            \n",
    "            self._get_dir_(output_intersection_dir)\n",
    "            \n",
    "            feature_intersection_df[[\"chrom\", \"start\", \"end\"]].to_csv(output_intersection_fname, sep=\"\\t\", index=False, header=False)\n",
    "\n",
    "    def _import_gene_counts_(self):\n",
    "        self.gene_counts = pd.read_csv(self.gene_expression, sep=\"\\t\", header=0, names=[\"GeneName\", \"Expression\"])\n",
    "    \n",
    "    def _map_feature_counts_to_genes_(self):\n",
    "        self.gene_counts[\"enhancer_count\"] = self.gene_counts[\"GeneName\"].map(self.df.groupby([\"baitName\"]).count()[\"baitChr\"])\n",
    "        \n",
    "        for _, tag in self.features_to_count.items():\n",
    "            self.gene_counts[f\"{tag}_count\"] = self.gene_counts[\"GeneName\"].map(self.df.groupby([\"baitName\"]).sum()[tag])\n",
    "    \n",
    "\n",
    "    def _get_dir_(self, dir: str, permissions=0o0775, exist_ok : bool=True):\n",
    "        \"\"\"Makes a directory at the given location\n",
    "        Args:\n",
    "            dir (str): Path of the directory\n",
    "            permissions ([type], optional): Permissions of directory. Defaults to 0o0775.\n",
    "            exist_ok (bool, optional): If True, program will continue if directory exists. Defaults to True.\n",
    "        Returns:\n",
    "            str: Absolute path to the created directory\n",
    "            \n",
    "        Example:\n",
    "        \n",
    "        >>> output_dir = get_dir(\"./output/\")\n",
    "        \"\"\"\n",
    "        try:\n",
    "            makedirs(dir, mode=permissions)\n",
    "        except error:\n",
    "            if not exist_ok:\n",
    "                raise\n",
    "\n",
    "    def _filter_expression_(self):\n",
    "        if self.nonzero_expression:\n",
    "            self.gene_counts = self.gene_counts[self.gene_counts[\"Expression\"] > 0]\n",
    "\n",
    "        if self.dropna_expression:\n",
    "            self.gene_counts = self.gene_counts.dropna()\n",
    "\n",
    "        self.gene_counts[\"GeneName_MeanExpression\"] = self.gene_counts[\"GeneName\"] \\\n",
    "            + \" \" + self.gene_counts[\"Expression\"].apply(str)\n",
    "        \n",
    "    def _calculate_spearman_(self):\n",
    "        self.corr = []\n",
    "        output_analysis_dir = os.path.join(self.output_dir, \"correlation_analysis\")\n",
    "\n",
    "        output_analysis_fname = os.path.join(output_analysis_dir, f\"{self.basename}_correlation_stats.tsv\")\n",
    "             \n",
    "        for _, tag in self.features_to_count.items():\n",
    "            tmp_df = self.gene_counts[[f\"{tag}_count\", \"Expression\"]].copy()\n",
    "            \n",
    "            s_corr, s_pval = spearmanr(tmp_df[f\"{tag}_count\"], tmp_df[\"Expression\"])\n",
    "            \n",
    "            p_corr, p_pval = pearsonr(tmp_df[f\"{tag}_count\"], tmp_df[\"Expression\"])\n",
    "\n",
    "            self.corr.append([s_corr, s_pval, p_corr, p_pval, tag])\n",
    "\n",
    "        self.corr_df = pd.DataFrame(self.corr, columns=[\"Spearman_corr\", \"Spearman_pval\", \"Pearson_corr\", \"Pearson_pval\", \"Feature\"])\n",
    "            \n",
    "        self._get_dir_(output_analysis_dir)\n",
    "\n",
    "        self.corr_df.to_csv(output_analysis_fname, sep=\"\\t\", index=False)\n",
    "\n",
    "    def _get_PIR_count_v_mean_(self):\n",
    "        \"\"\"Create the dataframe of the number of features overlapping PIRs by mean gene expression\n",
    "        \n",
    "        First you groupby the feature counts column. Then you find all of the mean gene expression values\n",
    "        associated with the number of features overlapping a PIR. The output is a dataframe that can be plotted\n",
    "        \"\"\"\n",
    "    \n",
    "        for _, tag in self.features_to_count.items():\n",
    "            col_name = f\"{tag}_count\"\n",
    "            output_analysis_dir = os.path.join(self.output_dir, \"expression_analysis\")\n",
    "\n",
    "            output_analysis_fname = os.path.join(output_analysis_dir, f\"{self.basename}_{tag}.tsv\")\n",
    "            \n",
    "            # Create the dataframe\n",
    "            self.pir_count_v_mean = pd.DataFrame(self.gene_counts.groupby([col_name])[\"GeneName_MeanExpression\"].apply(list)).reset_index().explode(\"GeneName_MeanExpression\")\n",
    "        \n",
    "            self.pir_count_v_mean[col_name] = self.pir_count_v_mean[col_name].apply(int)\n",
    "            \n",
    "            self.pir_count_v_mean[[\"Gene_Name\", \"Mean_Gene_Expression\"]] = self.pir_count_v_mean[\"GeneName_MeanExpression\"].str.split(\" \", n=2, expand=True)\n",
    "            \n",
    "            self.pir_count_v_mean.drop(\"GeneName_MeanExpression\",axis=1, inplace=True)\n",
    "            \n",
    "            self.pir_count_v_mean[\"Mean_Gene_Expression\"] = self.pir_count_v_mean[\"Mean_Gene_Expression\"].apply(float)\n",
    "            \n",
    "            self._get_dir_(output_analysis_dir)\n",
    "\n",
    "            self._calculate_spearman_()\n",
    "            \n",
    "            self.pir_count_v_mean.to_csv(output_analysis_fname, sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting PIRs From CHiCAGO output file\n",
    "\n",
    "I wrote a python class object to work with the CHiCAGO output text file. This piece of code will perform filtering of specific types of interactions, like promoter-to-promoter, or trans-chromosomal interactions. \n",
    "\n",
    "Example:\n",
    "\n",
    "```python\n",
    "input_file = \"../data/CHICAGO/hg38/inputs/ILC_5kb_within_newbmap_CHiCAGO_ABC_peakm.txt\"\n",
    "\n",
    "ILC3_data = ChicagoData(input_file)\n",
    "\n",
    "ILC3_data = ChicagoData(input_file, \n",
    "                        drop_off_target_bait=True, \n",
    "                        drop_off_target_oe=False, \n",
    "                        drop_trans_chrom=True,\n",
    "                        score_col=\"merged_score\",\n",
    "                        score_val=5,\n",
    "                        remove_p2p=True)\n",
    "                        \n",
    "ILC3_data.pir_df\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input Chicago File\n",
    "input_file = \"../data/CHICAGO/hg38/inputs/ILC_5kb_within_newbmap_CHiCAGO_ABC_peakm.txt.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/ATAC/ILC3_ATAC_peaks.bed.gz : Column will be saved as ATAC\n",
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/CHIP/ILC3_H3K27ac_peaks.bed.gz : Column will be saved as H3K27ac\n",
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/CHIP/ILC3_H3K4me3_peaks.bed.gz : Column will be saved as H3K4me3\n",
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/RE/ILC3_RE.bed.gz : Column will be saved as RE\n"
     ]
    }
   ],
   "source": [
    "ILC3_data = ChicagoData(input_file, \n",
    "                        drop_off_target_bait=True, \n",
    "                        drop_off_target_oe=False, \n",
    "                        drop_trans_chrom=True,\n",
    "                        score_col=\"merged_score\",\n",
    "                        score_val=5,\n",
    "                        remove_p2p=True,\n",
    "                        features_to_count=ilc3_file_dict,\n",
    "                        gene_expression=\"/Users/caz3so/workspaces/tacazares/pchic/data/RNA/ILC3_mean_expression.tsv\",\n",
    "                        output_dir=\"/Users/caz3so/workspaces/tacazares/pchic/data/outputs\",\n",
    "                        output_basename=\"ILC_5kb_within_newbmap_CHiCAGO_ABC_peakm\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RE_count</th>\n",
       "      <th>Gene_Name</th>\n",
       "      <th>Mean_Gene_Expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>43.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>114.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>AADAT</td>\n",
       "      <td>28.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>AARD</td>\n",
       "      <td>412.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ABCA13</td>\n",
       "      <td>165.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>148</td>\n",
       "      <td>TTF2</td>\n",
       "      <td>1019.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>152</td>\n",
       "      <td>CD44</td>\n",
       "      <td>24542.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>152</td>\n",
       "      <td>NSMCE1</td>\n",
       "      <td>1461.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>320</td>\n",
       "      <td>ID2</td>\n",
       "      <td>2022.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>361</td>\n",
       "      <td>FYN</td>\n",
       "      <td>5506.75</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9811 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     RE_count Gene_Name  Mean_Gene_Expression\n",
       "0           0      A1CF                 43.00\n",
       "0           0     A2MP1                114.25\n",
       "0           0     AADAT                 28.25\n",
       "0           0      AARD                412.25\n",
       "0           0    ABCA13                165.75\n",
       "..        ...       ...                   ...\n",
       "98        148      TTF2               1019.25\n",
       "99        152      CD44              24542.50\n",
       "99        152    NSMCE1               1461.50\n",
       "100       320       ID2               2022.00\n",
       "101       361       FYN               5506.75\n",
       "\n",
       "[9811 rows x 3 columns]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ILC3_data.pir_count_v_mean"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Processing Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/caz3so/workspaces/tacazares/pchic/notebooks'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "ilc3_file = \"../data/CHICAGO/hg38/inputs/ILC_5kb_within_newbmap_CHiCAGO_ABC_peakm.txt.gz\"\n",
    "ilc3_old_file = \"../data/CHICAGO/hg38/inputs/ILC3_merged_bin5K_score5.txt.gz\"\n",
    "\n",
    "cd4_1M_old_file = \"../data/CHICAGO/hg38/inputs/CD4_1M_50K_merged_reweighting_peakmatrix_score5.txt.gz\"\n",
    "cd4_file = \"../data/CHICAGO/hg38/inputs/CD4_1M_50K_5kb_within_newbmap_CHiCAGO_peakm.txt.gz\"\n",
    "cd4_ABC = \"../data/CHICAGO/hg38/inputs/CD4_1M_50K_5kb_within_newbmap_CHiCAGO_ABC_peakm.txt.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "ilc3_file_dict = {\"/Users/caz3so/workspaces/tacazares/pchic/data/peaks/ATAC/ILC3_ATAC_peaks.bed.gz\": \"ATAC\",\n",
    "                  \"/Users/caz3so/workspaces/tacazares/pchic/data/peaks/CHIP/ILC3_H3K27ac_peaks.bed.gz\": \"H3K27ac\",\n",
    "                  \"/Users/caz3so/workspaces/tacazares/pchic/data/peaks/CHIP/ILC3_H3K4me3_peaks.bed.gz\": \"H3K4me3\",\n",
    "                  \"/Users/caz3so/workspaces/tacazares/pchic/data/peaks/RE/ILC3_RE.bed.gz\": \"RE\"}\n",
    "\n",
    "cd4_file_dict = {\"/Users/caz3so/workspaces/tacazares/pchic/data/peaks/ATAC/CD4_ATAC_peaks.bed.gz\": \"ATAC\",\n",
    "                 \"/Users/caz3so/workspaces/tacazares/pchic/data/peaks/CHIP/S008H1H1.ERX547940.H3K27ac.bwa.GRCh38.20150527.bed.gz\": \"H3K27ac\",\n",
    "                 \"/Users/caz3so/workspaces/tacazares/pchic/data/peaks/CHIP/S008H1H1.ERX547958.H3K4me3.bwa.GRCh38.20150527.bed.gz\": \"H3K4me3\",\n",
    "                 \"/Users/caz3so/workspaces/tacazares/pchic/data/peaks/RE/CD4_RE.bed.gz\": \"RE\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "ilc_gene_expression = \"/Users/caz3so/workspaces/tacazares/pchic/data/RNA/ILC3_mean_expression.tsv\"\n",
    "cd4_gene_expression = \"/Users/caz3so/workspaces/tacazares/pchic/data/RNA/CD4_mean_expression.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir = \"../data/outputs\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/peaks/ATAC/ILC3_ATAC_peaks.bed.gz : Column will be saved as ATAC\n",
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/peaks/CHIP/ILC3_H3K27ac_peaks.bed.gz : Column will be saved as H3K27ac\n",
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/peaks/CHIP/ILC3_H3K4me3_peaks.bed.gz : Column will be saved as H3K4me3\n",
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/peaks/RE/ILC3_RE.bed.gz : Column will be saved as RE\n",
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/peaks/ATAC/CD4_ATAC_peaks.bed.gz : Column will be saved as ATAC\n",
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/peaks/CHIP/S008H1H1.ERX547940.H3K27ac.bwa.GRCh38.20150527.bed.gz : Column will be saved as H3K27ac\n",
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/peaks/CHIP/S008H1H1.ERX547958.H3K4me3.bwa.GRCh38.20150527.bed.gz : Column will be saved as H3K4me3\n",
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/peaks/RE/CD4_RE.bed.gz : Column will be saved as RE\n",
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/peaks/ATAC/CD4_ATAC_peaks.bed.gz : Column will be saved as ATAC\n",
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/peaks/CHIP/S008H1H1.ERX547940.H3K27ac.bwa.GRCh38.20150527.bed.gz : Column will be saved as H3K27ac\n",
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/peaks/CHIP/S008H1H1.ERX547958.H3K4me3.bwa.GRCh38.20150527.bed.gz : Column will be saved as H3K4me3\n",
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/peaks/RE/CD4_RE.bed.gz : Column will be saved as RE\n",
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/peaks/ATAC/ILC3_ATAC_peaks.bed.gz : Column will be saved as ATAC\n",
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/peaks/CHIP/ILC3_H3K27ac_peaks.bed.gz : Column will be saved as H3K27ac\n",
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/peaks/CHIP/ILC3_H3K4me3_peaks.bed.gz : Column will be saved as H3K4me3\n",
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/peaks/RE/ILC3_RE.bed.gz : Column will be saved as RE\n",
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/peaks/ATAC/CD4_ATAC_peaks.bed.gz : Column will be saved as ATAC\n",
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/peaks/CHIP/S008H1H1.ERX547940.H3K27ac.bwa.GRCh38.20150527.bed.gz : Column will be saved as H3K27ac\n",
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/peaks/CHIP/S008H1H1.ERX547958.H3K4me3.bwa.GRCh38.20150527.bed.gz : Column will be saved as H3K4me3\n",
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/peaks/RE/CD4_RE.bed.gz : Column will be saved as RE\n"
     ]
    }
   ],
   "source": [
    "# Import ILC3 data\n",
    "ilc3 = ChicagoData(ilc3_file, \n",
    "                        drop_off_target_bait=True, \n",
    "                        drop_off_target_oe=False, \n",
    "                        drop_trans_chrom=True,\n",
    "                        score_col=\"merged_score\",\n",
    "                        score_val=5,\n",
    "                        remove_p2p=True,\n",
    "                        features_to_count=ilc3_file_dict,\n",
    "                        gene_expression=ilc_gene_expression,\n",
    "                        output_dir=out_dir,\n",
    "                        output_basename=\"ILC_5kb_within_newbmap_CHiCAGO_ABC_peakm\")\n",
    "\n",
    "# Import cd4 file CD4_1M_50K_5kb_within_newbmap_CHiCAGO_peakm.txt\n",
    "cd4 = ChicagoData(cd4_file, \n",
    "                        drop_off_target_bait=True, \n",
    "                        drop_off_target_oe=False, \n",
    "                        drop_trans_chrom=True,\n",
    "                        score_col=\"merged_score\",\n",
    "                        score_val=5,\n",
    "                        remove_p2p=True,\n",
    "                        features_to_count=cd4_file_dict,\n",
    "                        gene_expression=cd4_gene_expression,\n",
    "                        output_dir=out_dir,\n",
    "                        output_basename=\"CD4_1M_50K_5kb_within_newbmap_CHiCAGO_peakm\")\n",
    "\n",
    "# Import cd4_abc file CD4_1M_50K_5kb_within_newbmap_CHiCAGO_ABC_peakm.txt\n",
    "cd4_abc = ChicagoData(cd4_ABC, \n",
    "                        drop_off_target_bait=True, \n",
    "                        drop_off_target_oe=False, \n",
    "                        drop_trans_chrom=True,\n",
    "                        score_col=\"merged_score\",\n",
    "                        score_val=5,\n",
    "                        remove_p2p=True,\n",
    "                        features_to_count=cd4_file_dict,\n",
    "                        gene_expression=cd4_gene_expression,\n",
    "                        output_dir=out_dir,\n",
    "                        output_basename=\"CD4_1M_50K_5kb_within_newbmap_CHiCAGO_ABC_peakm\")\n",
    "\n",
    "# Import old ILC3 data\n",
    "ilc3_old = ChicagoData(ilc3_old_file, \n",
    "                        drop_off_target_bait=True, \n",
    "                        drop_off_target_oe=False, \n",
    "                        drop_trans_chrom=True,\n",
    "                        score_col=\"hILC3_merged_bin5k\",\n",
    "                        score_val=5,\n",
    "                        remove_p2p=True,\n",
    "                        features_to_count=ilc3_file_dict,\n",
    "                        gene_expression=ilc_gene_expression,\n",
    "                        output_dir=out_dir,\n",
    "                        output_basename=\"ILC3_merged_bin5K_score5\")\n",
    "\n",
    "# Import old CD4_1M data\n",
    "cd4_1M_old = ChicagoData(cd4_1M_old_file, \n",
    "                        drop_off_target_bait=True, \n",
    "                        drop_off_target_oe=False, \n",
    "                        drop_trans_chrom=True,\n",
    "                        score_col=\"M1\",\n",
    "                        score_val=5,\n",
    "                        remove_p2p=True,\n",
    "                        features_to_count=cd4_file_dict,\n",
    "                        gene_expression=cd4_gene_expression,\n",
    "                        output_dir=out_dir,\n",
    "                        output_basename=\"CD4_1M_50K_merged_reweighting_peakmatrix_score5\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Spearman_corr</th>\n",
       "      <th>Spearman_pval</th>\n",
       "      <th>Pearson_corr</th>\n",
       "      <th>Pearson_pval</th>\n",
       "      <th>Feature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.227638</td>\n",
       "      <td>2.493171e-67</td>\n",
       "      <td>0.039472</td>\n",
       "      <td>0.003000</td>\n",
       "      <td>ATAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.114975</td>\n",
       "      <td>4.326785e-18</td>\n",
       "      <td>0.024818</td>\n",
       "      <td>0.062110</td>\n",
       "      <td>H3K27ac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.167423</td>\n",
       "      <td>8.403177e-37</td>\n",
       "      <td>0.028969</td>\n",
       "      <td>0.029429</td>\n",
       "      <td>H3K4me3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.229787</td>\n",
       "      <td>1.318676e-68</td>\n",
       "      <td>0.039183</td>\n",
       "      <td>0.003219</td>\n",
       "      <td>RE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Spearman_corr  Spearman_pval  Pearson_corr  Pearson_pval  Feature\n",
       "0       0.227638   2.493171e-67      0.039472      0.003000     ATAC\n",
       "1       0.114975   4.326785e-18      0.024818      0.062110  H3K27ac\n",
       "2       0.167423   8.403177e-37      0.028969      0.029429  H3K4me3\n",
       "3       0.229787   1.318676e-68      0.039183      0.003219       RE"
      ]
     },
     "execution_count": 178,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cd4_1M_old.corr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "c40e1b489fb11c0bd8cc5a783b5d3b77609a7988a043d25bde579eefe634d11a"
  },
  "kernelspec": {
   "display_name": "Python 3.9.10 ('maxatac')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
