{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from scipy.stats import pearsonr\n",
    "from scipy.stats import spearmanr\n",
    "import pybedtools\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ModExpressionAnalysis(object):\n",
    "    \"\"\"Import CHiCAGO data\n",
    "    \"\"\"\n",
    "    def __init__(self,\n",
    "                 filename: str,\n",
    "                 features_to_count: dict = {},\n",
    "                 gene_expression: str = \"\",\n",
    "                 nonzero_expression: bool = True,\n",
    "                 dropna_expression: bool = True,\n",
    "                 output_dir: str = \"\",\n",
    "                 output_basename: str = \"\"\n",
    "                 ):\n",
    "        \"\"\"Initialize the object\n",
    "\n",
    "        Args:\n",
    "            filename (str): Input CHICAGO txt file\n",
    "            drop_off_target (bool, optional): Drop off target interactions. Defaults to =True.\n",
    "            drop_trans_chrom (bool, optional): Drop transchromosomal interactions. Defaults to =True.\n",
    "        \"\"\"\n",
    "        # Set filename to the input filename\n",
    "        self.filename = filename\n",
    "        # Map feature counts to PIR if provided\n",
    "        self.features_to_count = features_to_count\n",
    "        # Import the gene expression matrix\n",
    "        self.gene_expression = gene_expression\n",
    "        # Only keep non-zero expression\n",
    "        self.nonzero_expression = nonzero_expression\n",
    "        # Drop na Values\n",
    "        self.dropna_expression = dropna_expression\n",
    "        # Set the output directory\n",
    "        self.output_dir = output_dir\n",
    "        # Set the output basename \n",
    "        self.basename = output_basename\n",
    "        \n",
    "        # Read file into DF\n",
    "        self._read_file_()\n",
    "        \n",
    "        # Get the PIR df\n",
    "        self._get_PIR_df_()\n",
    "        \n",
    "        self._get_feature_counts_()\n",
    "        \n",
    "        self._import_gene_counts_()\n",
    "            \n",
    "        self._map_feature_counts_to_genes_()\n",
    "        \n",
    "        self._map_ABC_counts_to_genes_()\n",
    "        \n",
    "        self._filter_expression_()\n",
    "        \n",
    "        self._get_PIR_count_v_mean_()\n",
    "                            \n",
    "    def _read_file_(self):\n",
    "        \"\"\"Read in original file\n",
    "        \"\"\"\n",
    "        # Read in original file and save\n",
    "        self.df =  pd.read_csv(self.filename, sep=\"\\t\", header=0, low_memory=False)\n",
    "        \n",
    "        self.df[\"ABC_label\"] = self.df[\"ABC.Score\"].apply(lambda x: \"ABC\" if x == 0 else \"CHiCAGO\")\n",
    "    \n",
    "    def _get_PIR_df_(self):\n",
    "        \"\"\"Get a DF of all PIR interactions\n",
    "        \"\"\"\n",
    "        self.pir_df = self.df[[\"oeChr\", \"oeStart\", \"oeEnd\", \"oe_interval_ID\"]].drop_duplicates(subset=[\"oeChr\", \"oeStart\", \"oeEnd\"], keep=\"first\")\n",
    "        \n",
    "        self.PIR_bt = pybedtools.BedTool.from_dataframe(self.pir_df)\n",
    "\n",
    "    def _get_feature_counts_(self):\n",
    "        \"\"\"Get the counts of features that overlap PIRs and map them back to the pcHiC interaction\n",
    "        \"\"\"\n",
    "        for file, tag in self.features_to_count.items():\n",
    "            print(f\"Importing {file} : Column will be saved as {tag}\")\n",
    "            output_intersection_dir = os.path.join(self.output_dir, \"PIR_intersection\")\n",
    "\n",
    "            output_intersection_fname = os.path.join(output_intersection_dir, f\"{self.basename}_PIR_intersect_{tag}.bed\")\n",
    "            \n",
    "            # Import and convert the featuers to a bedtools\n",
    "            feature_bt = pybedtools.BedTool(file)\n",
    "            \n",
    "            # Intersect the features to get counts\n",
    "            feature_counts = self.PIR_bt.intersect(feature_bt, c=True)\n",
    "            \n",
    "            # Intersect the features to get overlaps (true intersections)\n",
    "            feature_intersection = self.PIR_bt.intersect(feature_bt)\n",
    "            \n",
    "            feature_intersection_sort = feature_intersection.sort()\n",
    "            \n",
    "            # Convert bedtools to pandas\n",
    "            feature_counts_df = feature_counts.to_dataframe()\n",
    "                        \n",
    "            # Convert bedtools to dataframe\n",
    "            feature_intersection_df = feature_intersection_sort.to_dataframe()\n",
    "                        \n",
    "            # Create a dictionary of counts \n",
    "            counts_dict =  pd.Series(feature_counts_df[\"score\"].values,index=feature_counts_df[\"name\"]).to_dict()\n",
    "\n",
    "            # Map the counts back to the CHICAGO dataframe\n",
    "            self.df[tag] = self.df[\"oe_interval_ID\"].map(counts_dict)\n",
    "                        \n",
    "            feature_intersection_df[[\"chrom\", \"start\", \"end\"]].to_csv(output_intersection_fname, sep=\"\\t\", index=False, header=False)\n",
    "\n",
    "    def _import_gene_counts_(self):\n",
    "        self.gene_counts = pd.read_csv(self.gene_expression, sep=\"\\t\", header=0, names=[\"GeneName\", \"Expression\"])\n",
    "    \n",
    "    def _map_feature_counts_to_genes_(self):        \n",
    "        for _, tag in self.features_to_count.items():\n",
    "            self.gene_counts[f\"{tag}_count\"] = self.gene_counts[\"GeneName\"].map(self.df.groupby([\"baitName\"]).sum()[tag])\n",
    "\n",
    "    def _map_ABC_counts_to_genes_(self):        \n",
    "        for _, tag in self.features_to_count.items():\n",
    "            tmp = self.df[self.df[\"ABC_label\"] == \"ABC\"]\n",
    "            \n",
    "            self.gene_counts[\"ABC_count\"] = self.gene_counts[\"GeneName\"].map(tmp.groupby([\"baitName\"]).sum()[tag])\n",
    "\n",
    "            self.gene_counts[\"CHiCAGO_count\"] = self.gene_counts[f\"{tag}_count\"] - self.gene_counts[\"ABC_count\"]\n",
    "\n",
    "    def _filter_expression_(self):\n",
    "        if self.nonzero_expression:\n",
    "            self.gene_counts = self.gene_counts[self.gene_counts[\"Expression\"] > 0]\n",
    "\n",
    "        if self.dropna_expression:\n",
    "            self.gene_counts = self.gene_counts.dropna()\n",
    "\n",
    "        self.gene_counts[\"GeneName_MeanExpression\"] = self.gene_counts[\"GeneName\"] \\\n",
    "            + \" \" + self.gene_counts[\"Expression\"].apply(str)\n",
    "        \n",
    "    def _get_PIR_count_v_mean_(self):\n",
    "        \"\"\"Create the dataframe of the number of features overlapping PIRs by mean gene expression\n",
    "        \n",
    "        First you groupby the feature counts column. Then you find all of the mean gene expression values\n",
    "        associated with the number of features overlapping a PIR. The output is a dataframe that can be plotted\n",
    "        \"\"\"\n",
    "    \n",
    "        for _, tag in self.features_to_count.items():\n",
    "            col_name = f\"{tag}_count\"\n",
    "            \n",
    "            # Create the dataframe\n",
    "            self.pir_count_v_mean = pd.DataFrame(self.gene_counts.groupby([col_name])[\"GeneName_MeanExpression\"].apply(list)).reset_index().explode(\"GeneName_MeanExpression\")\n",
    "        \n",
    "            self.pir_count_v_mean[col_name] = self.pir_count_v_mean[col_name].apply(int)\n",
    "            \n",
    "            self.pir_count_v_mean[[\"Gene_Name\", \"Mean_Gene_Expression\"]] = self.pir_count_v_mean[\"GeneName_MeanExpression\"].str.split(\" \", n=2, expand=True)\n",
    "            \n",
    "            self.pir_count_v_mean.drop(\"GeneName_MeanExpression\",axis=1, inplace=True)\n",
    "            \n",
    "            self.pir_count_v_mean[\"Mean_Gene_Expression\"] = self.pir_count_v_mean[\"Mean_Gene_Expression\"].apply(float)\n",
    "                        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "ilc3_file = \"../data/outputs/modified_chicago/ILC_5kb_within_newbmap_CHiCAGO_ABC_peakm_modified.tsv\"\n",
    "\n",
    "cd4_ABC = \"../data/outputs/modified_chicago/CD4_1M_50K_5kb_within_newbmap_CHiCAGO_ABC_peakm_modified.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "ilc3_file_dict = {\"/Users/caz3so/workspaces/tacazares/pchic/data/peaks/RE/ILC3_RE.bed\": \"RE\"}\n",
    "\n",
    "cd4_file_dict = {\"/Users/caz3so/workspaces/tacazares/pchic/data/peaks/RE/CD4_RE.bed\": \"RE\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "ilc_gene_expression = \"/Users/caz3so/workspaces/tacazares/pchic/data/RNA/ILC3_mean_expression.tsv\"\n",
    "cd4_gene_expression = \"/Users/caz3so/workspaces/tacazares/pchic/data/RNA/CD4_mean_expression.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir = \"../data/outputs\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/peaks/RE/ILC3_RE.bed : Column will be saved as RE\n",
      "Importing /Users/caz3so/workspaces/tacazares/pchic/data/peaks/RE/CD4_RE.bed : Column will be saved as RE\n"
     ]
    }
   ],
   "source": [
    "# Import ILC3 data\n",
    "ilc3 = ModExpressionAnalysis(ilc3_file, \n",
    "                        features_to_count=ilc3_file_dict,\n",
    "                        gene_expression=ilc_gene_expression,\n",
    "                        output_dir=out_dir,\n",
    "                        output_basename=\"ILC_5kb_within_newbmap_CHiCAGO_ABC_peakm\")\n",
    "\n",
    "# Import cd4_abc file CD4_1M_50K_5kb_within_newbmap_CHiCAGO_ABC_peakm.txt\n",
    "cd4 = ModExpressionAnalysis(cd4_ABC, \n",
    "                        features_to_count=cd4_file_dict,\n",
    "                        gene_expression=cd4_gene_expression,\n",
    "                        output_dir=out_dir,\n",
    "                        output_basename=\"CD4_1M_50K_5kb_within_newbmap_CHiCAGO_ABC_peakm\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneName</th>\n",
       "      <th>Expression</th>\n",
       "      <th>RE_count</th>\n",
       "      <th>GeneName_MeanExpression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>474.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>A1BG 474.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>5.166667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>A1CF 5.166666666666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2ML1</td>\n",
       "      <td>68.833333</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A2ML1 68.83333333333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A2MP1</td>\n",
       "      <td>4.833333</td>\n",
       "      <td>0.0</td>\n",
       "      <td>A2MP1 4.833333333333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>118.833333</td>\n",
       "      <td>2.0</td>\n",
       "      <td>A4GALT 118.83333333333331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22444</th>\n",
       "      <td>ZSWIM6</td>\n",
       "      <td>176.500000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ZSWIM6 176.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22449</th>\n",
       "      <td>ZW10</td>\n",
       "      <td>563.500000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ZW10 563.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22452</th>\n",
       "      <td>ZXDA</td>\n",
       "      <td>21.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ZXDA 21.666666666666668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22456</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>465.500000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ZYG11B 465.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22457</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>424.166667</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ZYX 424.1666666666667</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9916 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      GeneName  Expression  RE_count    GeneName_MeanExpression\n",
       "0         A1BG  474.000000       2.0                 A1BG 474.0\n",
       "1         A1CF    5.166667       0.0     A1CF 5.166666666666667\n",
       "3        A2ML1   68.833333       1.0    A2ML1 68.83333333333333\n",
       "4        A2MP1    4.833333       0.0    A2MP1 4.833333333333333\n",
       "5       A4GALT  118.833333       2.0  A4GALT 118.83333333333331\n",
       "...        ...         ...       ...                        ...\n",
       "22444   ZSWIM6  176.500000       2.0               ZSWIM6 176.5\n",
       "22449     ZW10  563.500000       3.0                 ZW10 563.5\n",
       "22452     ZXDA   21.666667       0.0    ZXDA 21.666666666666668\n",
       "22456   ZYG11B  465.500000       2.0               ZYG11B 465.5\n",
       "22457      ZYX  424.166667       4.0      ZYX 424.1666666666667\n",
       "\n",
       "[9916 rows x 4 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ilc3.gene_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.10 ('maxatac')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "c40e1b489fb11c0bd8cc5a783b5d3b77609a7988a043d25bde579eefe634d11a"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
